uk.co.compucorp.civicrm.hrcore/CRM/HRCore/Upgrader/Steps/1004.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+trait CRM_HRCore_Upgrader_Steps_1004 {
+
+  /**
+   * Upgrader to set 'Home' as default location and set all 'Home' addresses as
+   * primary.
+   *
+   * @return bool
+   */
+  public function upgrade_1004() {
+    $this->up1004_setHomeLocationAsDefault();
+    $this->up1004_setAllHomeAddressesAsPrimary();
+
+    return true;
+  }
+
+  private function up1004_setHomeLocationAsDefault() {
+    civicrm_api3('LocationType', 'get', [
+      'sequential' => 1,
+      'name' => 'Home',
+      'api.LocationType.create' => [
+        'id' => '$value.id',
+        'is_default' => 1,
+        'is_reserved' => 1,
+      ],
+    ]);
+  }
+
+  private function up1004_setAllHomeAddressesAsPrimary() {
+    $homeLocation = civicrm_api3('LocationType', 'getsingle', [
+      'name' => "Home",
+    ]);
+
+    $contactsWithNonPrimaryHomeAddress = civicrm_api3('Address', 'get', [
+      'sequential' => 1,
+      'return' => array('contact_id'),
+      'location_type_id' => $homeLocation['id'],
+      'is_primary' => 0,
+    ]);
+
+    $contactIDs = [];
+    foreach ($contactsWithNonPrimaryHomeAddress['values'] as $currentContact) {
+      $contactIDs[] = $currentContact['contact_id'];
+    }
+
+    if (!empty($contactIDs)) {
+      $query = "
+        UPDATE civicrm_address
+        SET is_primary = CASE WHEN location_type_id = {$homeLocation['id']} THEN 1 ELSE 0 END
+        WHERE contact_id IN (" . implode(',', $contactIDs) . ")
+      ";
```

Comment:
I could use the API, but I'd need two different calls, one to set 'Home' addresses as primary, the other one to set all the other addresses as not primary. I could also use a single call, kind of, by cycling through all contacts, making each call to set the is_primary flag to true and let the Address BAO handle the setting of all other addresses to false, but that would obviously take a lot more time - O(1) vs O(n).

On the other note, my idea was to set the is_primary flag to true for all contacts that had a 'Home' address that is not primary.  Only for those contacts should 'Home' addresses be set as primary, and other addresses be unset. To do that I need the contact's ID.